### ISSUE TITLE
Research: determine which regular expression engine to use

### ISSUE THREAD

**Issue by tangledhelix** [DistributedProofreaders/guiguts-py#55]
Python's base regex engine (in the `re` module) is "similar" to Perl PCRE but not exactly compatible. This leaves us with a choice to make, and some research to do, to inform the choice: should we make GG2 backward-compatible with PP'er workflows? Many PP'ers have carefully honed regexes used in their workflows. An argument could be made that the best choice is to make GG2 perfectly compatible with those workflows. But that must be counterbalanced against the amount of effort required to accomplish the backward compatibility, and ongoing maint, etc.

References / projects to review:

- [Wikipedia: Comparison of regular expression engines][1]
- [Python 3.10: re — Regular expression operations][2]
- [regex module - a backward-compatible evolution of Python's "re"][3]
- [python-pcre: Python bindings for PCRE regex engine][4]
- [StackOverflow: How can I use PCRE regexes from a python script?][5]

[1]: https://en.wikipedia.org/wiki/Comparison_of_regular_expression_engines
[2]: https://docs.python.org/3.10/library/re.html
[3]: https://github.com/mrabarnett/mrab-regex
[4]: https://github.com/awahlig/python-pcre
[5]: https://stackoverflow.com/questions/16940150/how-can-i-use-pcre-regexes-from-a-python-script

At a cursory glance, the [python-pcre][4] option seems "best" in terms of least disruptive to users, but I have no opinion yet about the future maintainability about any of these options. I will look at this, though of course I would welcome any opinions or input from anyone who has time to read / think about this.

**Comment by tangledhelix** [comment]
One additional thing I want to clarify - there may exist two use cases for regular expressions. Some regex will end up being used internally by the application, and that's not what I'm talking about here. I'm talking about regexes input by the user, e.g. for Find/Replace, and other operations. If the user isn't inputting the regex, then I think the `re` module is the simplest and best choice.

**Comment by windymilla** [comment]
Good question Dan.

Not sure about this, but have done a small amount of research. The case you are considering where the regexes are input by the user are mostly passed as arguments to the `search` method on the `Text` widget. This apparently accepts "Tcl-style" regexes, which, depending on how up-to-date documentation is, are either "a subset of python regexes", or maybe "Advanced REs".
I don't know that using `python-pcre` would affect that.
Here are a couple of possibly relevant pages:
https://tcl.tk/man/tcl8.6/TkCmd/text.htm#M151
https://tcl.tk/man/tcl8.6/TclCmd/re_syntax.htm#M4

In the Perl/Tk docs it says it uses "the rules for regular expressions (see the regexp command for details)." (https://metacpan.org/pod/Tk::Text#WIDGET-METHODS)
However, there is no link to a "regexp" command, nor could I find a Perl/Tk version of that, just the Tk one mentioned above. 

So, in summary, it's possible that the user-supplied regexps in GG1 are subject to the same limitations as those in GG2, assuming both of them get passed to `Text::search`. Some testing of more advanced stuff like look-ahead/behind, would probably determine if it's likely to be a problem.

I suspect that "most" user's regexps would continue to work, and that the work involved in providing (rarely needed) full support for

**Comment by tangledhelix** [comment]
Interesting. I hadn't considered that the regex engine might come from Tcl. I just assumed it came from Perl.

Some links:
- [Tcl regexp command](https://www.tcl-lang.org/man/tcl8.7/TclCmd/regexp.html)
- [re_syntax – Syntax of Tcl regular expressions](https://www.tcl-lang.org/man/tcl8.7/TclCmd/re_syntax.html)

The first is the actual Tcl command, which may not be of interest since we aren't writing Tcl code. I include it because it could be useful for regex testing (see below). Assuming you're right, and the Tcl engine is used, the second link is probably more useful.

If it wasn't so late I would check the "Help with regex" forum thread and verify Tcl regex works with some examples that have been discussed there... maybe a different day I'll find time to check into that.

Here's a sample script for testing a regex against a string (assuming tclsh is available). This is a straightforward regex and shows how to reference capture groups.

```bash
#!/usr/bin/env tclsh

regexp {([A-Za-z]*).([A-Za-z]*)} "Foo Bar" a b c
puts "Full Match: $a"
puts "Sub Match1: $b"
puts "Sub Match2: $c"
```

The above script should output this:

```text
Full Match: Foo Bar
Sub Match1: Foo
Sub Match2: Bar
```

**Comment by tangledhelix** [comment]
I looked at this again, as did @windymilla. We're fairly sure that Perl Guiguts is using Perl's regex, not Tcl's. If you test for example `\p{Punct}` in Guiguts (or Perl), it works, but in Tcl it results in a syntax error.

I also looked again at [python-pcre](https://github.com/awahlig/python-pcre) - I'm concerned about its apparent lack of maintenance. There have been no commits in 8 years, and issues have been open for several years. It doesn't seem like there's any activity after 2015. We may not want to build a reliance on it, unless we're also willing to take on maintenance effort for it / fork the project.

**Comment by windymilla** [comment]
Note: Quentin is trying out the `regex` module while working on the `pptxt` tool.

**Comment by windymilla** [comment]
Settled on `regex` module, at least for now. Just about to start work on Search/Replace.
